fendermint/app/src/ipc.rs

```diff
@@ -0,0 +1,149 @@
+use crate::BlockHeight;
+use anyhow::{anyhow, Context};
+use base64::Engine;
+use fendermint_storage::{Codec, Encode, KVRead, KVReadable, KVStore};
+use fendermint_vm_actor_interface::{ipc, system};
+use fendermint_vm_interpreter::fvm::state::{FvmExecState, FvmStateParams};
+use fendermint_vm_interpreter::fvm::FvmMessage;
+use fendermint_vm_ipc_actors::gateway_getter_facet;
+use fendermint_vm_topdown::convert::{DecodeFunctionReturn, EncodeWithSignature};
+use fendermint_vm_topdown::IPCParentFinality;
+use fvm::engine::MultiEngine;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::{BytesDe, RawBytes};
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use num_traits::Zero;
+use std::sync::Arc;
+
+/// Queries the LATEST COMMITTED parent finality from the storage
+pub struct ParentFinalityQuery<DB, SS, S>
+where
+    SS: Blockstore + 'static,
+    S: KVStore,
+{
+    /// Database backing all key-value operations.
+    db: Arc<DB>,
+    /// State store, backing all the smart contracts.
+    ///
+    /// Must be kept separate from storage that can be influenced by network operations such as Bitswap;
+    /// nodes must be able to run transactions deterministically. By contrast the Bitswap store should
+    /// be able to read its own storage area as well as state storage, to serve content from both.
+    state_store: Arc<SS>,
+    /// Wasm engine cache.
+    multi_engine: Arc<MultiEngine>,
+    /// Namespace to store app state.
+    namespace: S::Namespace,
+}
+
+impl<DB, SS, S> ParentFinalityQuery<DB, SS, S>
+where
+    S: KVStore
+        + Codec<AppState>
+        + Encode<AppStoreKey>
+        + Encode<BlockHeight>
+        + Codec<FvmStateParams>,
+    DB: KVReadable<S> + Clone + 'static,
+    SS: Blockstore + Clone + 'static,
+{
+    pub fn new(
+        db: Arc<DB>,
+        state_store: Arc<SS>,
+        multi_engine: Arc<MultiEngine>,
+        namespace: S::Namespace,
+    ) -> Self {
+        Self {
+            db,
+            state_store,
+            multi_engine,
+            namespace,
+        }
+    }
+
+    pub fn get_committed_finality(&self) -> anyhow::Result<IPCParentFinality> {
+        let app_state = self
+            .get_committed_state()?
+            .ok_or_else(|| anyhow!("cannot get app state"))?;
+        let block_height = app_state.block_height as ChainEpoch;
+        let state_params = app_state.state_params;
+
+        let mut exec_state = FvmExecState::new(
+            self.state_store.clone(),
+            self.multi_engine.as_ref(),
+            block_height,
+            state_params,
+        )
+        .context("error creating execution state")?;
+
+        let msg = get_parent_finality_to_fvm(block_height)?;
+        let data = exec_state
+            .execute_implicit(msg)?
+            .msg_receipt
+            .return_data
+            .to_vec();
+        let decoded = decode_fevm_invoke(data)?;
+        DecodeFunctionReturn::<IPCParentFinality>::decode(decoded)
```

Comment:
it will be default value set in the contract. I assume after genesis, the query function should return whatever the contract returns. If that's the case, it would return `{height: 0, hash: vec![0; 32]}`